go/dexop/internal/controller/client_controller.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2025 Rackspace Technology.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-logr/logr"
+	dexv1alpha1 "github.com/rackerlabs/understack/go/dexop/api/v1alpha1"
+	dexmgr "github.com/rackerlabs/understack/go/dexop/dex"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const dexFinalizer = "dex.rax.io/finalizer"
+
+// ClientReconciler reconciles a Client object
+type ClientReconciler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	DexManager *dexmgr.DexManager
+}
+
+// +kubebuilder:rbac:groups=dex.rax.io,resources=clients,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=dex.rax.io,resources=clients/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=dex.rax.io,resources=clients/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.0/pkg/reconcile
+func (r *ClientReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+	reqLogger := ctrl.Log.WithValues("client", req.NamespacedName)
+
+	clientSpec, err := r.getClientSpec(ctx, req.NamespacedName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// resource was deleted but it is being finalized
+	if clientSpec == nil {
+		return ctrl.Result{}, nil
+	}
+	reqLogger.Info("Reconciling Client")
+
+	deleteRequested, err := r.handleDeletion(ctx, clientSpec)
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if deleteRequested {
```

Comment:
nitpick: personally I prefer to multiple if instead of else if.
```go
if err != nil {
	return ctrl.Result{}, err
} 
if deleteRequested {
    return ctrl.Result{}, nil
}
```

Reply:
I generally prefer that too, cannot remember why I wrote it this way - fixed as per suggestion, thanks!